### ISSUE TITLE
Incorporates assumption about ROIs

### ISSUE THREAD

**Issue by DanNBullock** [brainlife/validator-neuro-rois#1]
https://github.com/brainlife/validator-neuro-rois/blob/298168d6edda6a2b73edb52f20056efc1553a38c/validate.py#L51

If I'm interpreting this line correctly, it seems that ROIs can only be validated if they are binarized.  However, it is conceivable that non-binarized ROIs might be generated for probabilistic masks, fMRI, or from tract density masks.  I was thinking that I could just convert those relevant datatypes to ROI and then pipe them through apps like the warp-to-atlas-space (https://brainlife.io/app/5c7319a5badd19003102e8d8), but it seems that this would run afoul of the validator.

It seems like there are a few options here:

- alter the ROI format to accept non-binarized, non-integer niftis
- create a new data format specific to these use-cases/data cases
- create app variants that are specific to the relevant datatypes (i.e. https://brainlife.io/datatype/592dded1436ee50ffd88f5d0)

**Comment by soichih** [comment]
Yes, the datatype is defined such that all ROIs should have 0/1 values. I think `merged_img` is used to construct a single thumbnail of all the rois. I don't think the validator prevents app developer from storing anything other than 0 or 1 though..

For probability map, I think @giulia-berto requested for this datatype (not used right now..)
> https://brainlife.io/datatype/60f69321b991113f21b4fdf6

Are you wanting to create something similar to neuro/probability-map but allows for multiple such maps? Maybe call it neuro/probability-maps (similar to tractmasks)?

**Comment by DanNBullock** [comment]
Hrm, well if the validator doesn't explicitly stop it, then there may not be a problem here, ultimately.  

For now I think my quick solution will just be to convert the tractDensity datatype to a ROI, and go from there.

Conceptually, this is like a probability-map, but I wouldn't assume that all non-integer niftis are probability maps.  There could be other float-storage niftis out there that aren't tracking probability.

**Comment by soichih** [comment]
We might very well update our validator so that only 0/1 are allowed for rois

**Comment by DanNBullock** [comment]
So... I should do this quick before you implement that change?  : )

**Comment by giulia-berto** [comment]
The datatype probability-map
https://brainlife.io/datatype/60f69321b991113f21b4fdf6 was meant to store
continuous values between 0 and 1. Binary values are meant to be used only
for the mask datatypes, right? Like masks or tractmasks.

**Comment by giulia-berto** [comment]
Ah sorry, I realized now that you were talking about the neuro/rois
datatype. Yes, I think the validator should be updated to take only binary
values in that case.

**Comment by soichih** [comment]
@DanNBullock 

>  I should do this 

What do you mean by "this"? Do you want to create a new datatype still?

**Comment by DanNBullock** [comment]
My proximal goal was just to run this the [warp to MNI app](https://brainlife.io/app/5c7319a5badd19003102e8d8) on the output from the [tract density map creation app](https://brainlife.io/app/6068bf10c7f80a3c109411b0).  However, the latter app outputs tractmasks instead of ROIs (which are required by the warp app) so, as it stands, I cannot warp these masks to mni space.

To my understanding the possible resolutions of this were

- Converting to ROI datatype (now suggested to be ill advised by our preceding conversation) 
- Create a new version of the warp app to deal with tractmasks (I kind of don't like this idea just because I am really averse to creating lots of app variants that are all essentially doing the same thing--it creates a fairly large maintenance burden).
- Create some more omnibus datatype that encompasses tractmasks, but also other datatypes that are essentially the same, convert that datatype and create a new app for that (this option still has the same problem as the previous one--in that it creates a new app--but may be wiser in the long run as it will preclude additional maintenance as we expand our other datatype options.

I'm neutral on the creation of a new datatype, but here's some of my thoughts on that possibility:

- **ROIs**:  will be limited to binarized niftis; a directory of these.  In essence a collection of masks
- **probability-map**: (as @giulia-berto pointed out) these are nifitis that store (presumably) floats, but all values are between 0 and 1
-  **Parcellation/volume**:  For the moment these are nifitis with whole numbers (int?) and a corresponding key/dictionary indicating the identity of the numbers.  In essence this is largely functionally equivalent to **ROIs** __except__ that **ROIs** could allow for the same voxel to have different labels across the different files.
- **tractmasks**:  not sure how well specified this is, but inherent in the "masks" component of this name, it seems the intent is for this to be binarized (whether or not this is enforced is another matter).  Also, the plural here indicates that this is a directory of such masks.
- **various other nifti types**:  If I am not mistaken, other nifiti datatypes (T1,T2,DWI) may contain floats.

The (minimally specified) data use case that I would be considering in this case:

The files that I am creating with the [tract density app](https://brainlife.io/app/6068bf10c7f80a3c109411b0) be a folder of nifits, all corresponding to the same reference space/anatomy, containing float values, not bounded between 0 and 1.  So it's kind of a mishmash of existing cases.  I should also add that I suspect there are other data cases / outputs that would be characterized by this same structuring **and** warrant access to the same categories of apps.

More broadly, I feel that this (and the proliferation of nifit-based datatypes) highlights the need for a more taxonomic (and less piecemeal) approach to managing nifti datatypes.  There are a finite number variables/domains across which they can exhibit categorical variance, so it should be possible to create some systematic (i.e. machine readable/parsable) method for handling/routing these.  I kind of touch on this issue in a few other issues I've created ([Cross-datatype datatype tags](https://github.com/brainlife/brainlife/issues/85); [Mandatory / automatic assignment of some datatype tags](https://github.com/brainlife/brainlife/issues/86); [Accept .gz or .nii upload input, interconvert if necessary](https://github.com/brainlife/brainlife/issues/81)), but I recognize that the actual implementation of such a system would likely be quite labor-intensive/difficult.


@soichih I guess a short term solution is to just create a new app that takes in tractmask and outputs a tractmask?  This still leaves open the question of whether we want tractmask to actually include nifti files with non-binary data, but it may be a short term solution.  Otherwise it looks like we would need to create a new datatype (again, something that I'm kind of loathe to do due to the increased maintenance burden of proliferating types like this and the certainty of later-date changes to the datatype requiring retooling of all impacted apps).

**Comment by soichih** [comment]
It looks like we should probably try to deprecate neuro/tractmasks in favor of neuro/rois. But that's another story..

My understanding is that, both neuro/tractmasks and neuro/rois are binarized, so you must not use them for the purpose of storing tract densities.

The course of action that I recommend is the following.

1) register a new datatype that can store *multile* non-binalized nifti files. I am somewhat against creating a "generic" multi-nifties datatype (with "density-map" as a tag). We don't create datatype based on data format. So, in our case, creating "neuro/probability-densities" makes sense. I just registered this just to get things going. https://brainlife.io/datatype/61151f8ca5a04c1ad3f9a2d1

2) Update @DanNBullock 's "tract density masks" to use the new neuro/probability-densitites datatype (update app to output files under `/densitites` instead of `/masks`.

3) Update brainlife/app-subj2reference so that it can handle neuro/probability-densities. I believe the existing code should "just work". We just need to register a new App that passes neuro/probability-densities as "rois" as expected in the output and treat the output "rois" as output neuro/probability-densities.

Please let me know if this approach makes sense!

**Comment by francopestilli** [comment]
**This makes sense to me too**
> 1. register a new datatype that can store _multile_ non-binalized nifti files. I am somewhat against creating a "generic" multi-nifties datatype (with "density-map" as a tag). We don't create datatype based on data format. So, in our case, creating "neuro/probability-densities" makes sense. I just registered this just to get things going. https://brainlife.io/datatype/61151f8ca5a04c1ad3f9a2d1

**please let's do this**
> 2. Update @DanNBullock 's "tract density masks" to use the new neuro/probability-densitites datatype (update app to output files under `/densitites` instead of `/masks`.

**agreed**
> 3. Update brainlife/app-subj2reference so that it can handle neuro/probability-densities. I believe the existing code should "just work". We just need to register a new App that passes neuro/probability-densities as "rois" as expected in the output and treat the output "rois" as output neuro/probability-densities.

@DanNBullock @soichih

**Comment by DanNBullock** [comment]
> 3.  Update brainlife/app-subj2reference so that it can handle neuro/probability-densities. I believe the existing code should "just work". We just need to register a new App that passes neuro/probability-densities as "rois" as expected in the output and treat the output "rois" as output neuro/probability-densities.

Does this entail creating a new fork of the repo or modifying the existing one?  I think there are certian pipelines (that have been used in publications) that are set up for ROIs datatype.  Otherwise, a new app instance seems fine.

> 1.  register a new datatype that can store multiple non-binalized nifti files. I am somewhat against creating a "generic" multi-nifties datatype (with "density-map" as a tag). We don't create datatype based on data format. So, in our case, creating "neuro/probability-densities" makes sense. I just registered this just to get things going. https://brainlife.io/datatype/61151f8ca5a04c1ad3f9a2d1

I think this is close to being on target, however as written, its indicating that this datatype is 0-1 bounded, which is not accurate.  The files that my app are outputting aren't "probabilities" _per-se_ as they can simply be counts of the number of streamlines in a voxel.  True, it can also be normalized which would bring it into alignment with a 0-1 bound, but it doesn't default to this.

Edit: Maybe I'm misinterpreting the name of this and the "-" is indicating that the data content can be **either** probabilities **or** densities.  Was that the intent?

>  2. Update @DanNBullock 's "tract density masks" to use the new neuro/probability-densitites datatype (update app to output files under /densitites instead of /masks.

I can go ahead and do this, as I don't think anyone else is using this app right now.  I've deleted the project contents that related to the old version of this, as the old outputs couldn't be used for anything in light of our proposed changes.

**Comment by soichih** [comment]
> Does this entail creating a new fork of the repo or modifying the existing one? 

No, I am hoping that you can just use the same code / repo. I don't think `antsApplyTransforms` cares what datatype the nifti stores. I just registered this App > https://brainlife.io/app/61158b45a5a04c9673f9d510 As you can see, the only thing I changed is the input/output datatype. It uses the same config.json input key and output directory so the same code *should* work..

> True, it can also be normalized which would bring it into alignment with a 0-1 bound, but it doesn't default to this.

Yes, you will need to normalize so the value will be between 0-1 (thus "probability"). Similar to https://brainlife.io/datatype/60f69321b991113f21b4fdf6 

@giulia-berto Does probability-map datatype contains values in 0.0 - 0.1 range, correct? 

> Edit: Maybe I'm misinterpreting the name of this and the "-" is indicating that the data content can be either probabilities or densities. Was that the intent?

I will update it to just "neuro/probabilities".

**Comment by DanNBullock** [comment]
In this post (https://github.com/brainlife/validator-neuro-rois/issues/1#issuecomment-895556105) she indicate it was bound in this way.

**Comment by giulia-berto** [comment]
neuro/probability-map datatype contains values in 0.0 - 1.0 range (only one
nifti file). Do we want to keep both the datatypes neuro/probability-map
and neuro/probability?

**Comment by DanNBullock** [comment]
Ah, I forgot to add that the output of [the app in question](https://brainlife.io/app/6068bf10c7f80a3c109411b0), has multiple niftis--one for every white matter tract in the input WMC object.
